File: LeetCode/125.ValidPalindrome.cpp
//https://leetcode.com/problems/valid-palindrome/
// two pointer
class Solution {
public:
    bool isPalindrome(string s) {
        int start = 0, end = s.size() - 1;
        while(start < end){
        while(start < end && !isalnum(s[start])){
            start++;
        }
                  
          while(start < end && !isalnum(s[end])){
              end--;
          }
              
          if(start < end && tolower(s[start++]) != tolower(s[end--])){
              return false;
          }
        }
        return true;
      }
};

File: LeetCode/Python3/125.ValidPalindrome.py
# https://leetcode.com/problems/valid-palindrome/submissions/
# Tow pointer
# Space Complexity O(1)
# Time Complexity O(n)


class Solution:
    def isPalindrome(self, s: str) -> bool:
        l, r = 0, len(s) - 1
        while l < r:
            while l < r and not s[l].isalnum():
                l += 1
            while l < r and not s[r].isalnum():
                r -= 1

            if s[l].lower() != s[r].lower():
                return False
            l = l+1
            r = r-1
        return True
